File: apps/api/src/domain/entities/knowledge.py
from dataclasses import dataclass, field
from datetime import UTC, datetime
from uuid import UUID, uuid4


@dataclass
class KnowledgeBase:
    name: str
    created_by: UUID
    id: UUID = field(default_factory=uuid4)
    description: str | None = None
    institution: str | None = None
    subject: str | None = None
    language: str = "en"
    is_active: bool = True
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))


@dataclass
class Document:
    knowledge_base_id: UUID
    uploaded_by: UUID
    filename: str
    file_type: str
    file_size: int
    id: UUID = field(default_factory=uuid4)
    title: str | None = None
    description: str | None = None
    storage_path: str = ""
    status: str = "pending"        # pending | processing | ready | failed
    chunk_count: int = 0
    error_message: str | None = None
    metadata: dict = field(default_factory=dict)
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    def mark_processing(self) -> None:
        self.status = "processing"
        self.updated_at = datetime.now(UTC)

    def mark_ready(self, chunk_count: int) -> None:
        self.status = "ready"
        self.chunk_count = chunk_count
        self.updated_at = datetime.now(UTC)

    def mark_failed(self, error: str) -> None:
        self.status = "failed"
        self.error_message = error
        self.updated_at = datetime.now(UTC)


@dataclass
class DocumentChunk:
    document_id: UUID
    knowledge_base_id: UUID
    content: str
    chunk_index: int
    id: UUID = field(default_factory=uuid4)
    vector_id: str = ""             # ID of the corresponding point in Qdrant
    token_count: int = 0
    page_number: int | None = None
    char_start: int = 0
    char_end: int = 0
    metadata: dict = field(default_factory=dict)
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))


@dataclass
class IngestionJob:
    document_id: UUID
    id: UUID = field(default_factory=uuid4)
    status: str = "queued"         # queued | parsing | chunking | embedding | indexing | done | failed
    progress: int = 0              # 0–100
    error_message: str | None = None
    created_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    updated_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    def advance(self, status: str, progress: int) -> None:
        self.status = status
        self.progress = progress
        self.updated_at = datetime.now(UTC)

    def fail(self, error: str) -> None:
        self.status = "failed"
        self.error_message = error
        self.updated_at = datetime.now(UTC)

    def complete(self) -> None:
        self.status = "done"
        self.progress = 100
        self.updated_at = datetime.now(UTC)

File: client/apps/web/lib/api/knowledge.ts
import { apiClient } from './client'

export interface KnowledgeBase {
  id: string
  name: string
  description: string | null
  institution: string | null
  subject: string | null
  language: string
  is_active: boolean
  document_count: number
  created_at: string
  updated_at: string
}

export interface Document {
  id: string
  knowledge_base_id: string
  filename: string
  title: string | null
  file_type: string
  file_size: number
  status: 'pending' | 'processing' | 'ready' | 'failed'
  chunk_count: number
  error_message: string | null
  created_at: string
  updated_at: string
}

export interface Chunk {
  id: string
  document_id: string
  chunk_index: number
  content: string
  token_count: number
  page_number: number | null
  metadata: Record<string, unknown>
}

export interface IngestionJob {
  id: string
  document_id: string
  status: string
  progress: number
  error_message: string | null
  created_at: string
  updated_at: string
}

export const knowledgeApi = {
  listKnowledgeBases: (page = 1, limit = 20) =>
    apiClient.get<KnowledgeBase[]>(`/api/v1/library?page=${page}&limit=${limit}`),

  createKnowledgeBase: (data: { name: string; description?: string; institution?: string; subject?: string }) =>
    apiClient.post<KnowledgeBase>('/api/v1/library', data),

  getKnowledgeBase: (id: string) => apiClient.get<KnowledgeBase>(`/api/v1/library/${id}`),

  deleteKnowledgeBase: (id: string) => apiClient.delete<void>(`/api/v1/library/${id}`),

  listDocuments: (kbId: string, page = 1, limit = 20) =>
    apiClient.get<Document[]>(`/api/v1/documents/${kbId}?page=${page}&limit=${limit}`),

  getDocument: (kbId: string, docId: string) =>
    apiClient.get<Document>(`/api/v1/documents/${kbId}/${docId}`),

  deleteDocument: (kbId: string, docId: string) =>
    apiClient.delete<void>(`/api/v1/documents/${kbId}/${docId}`),

  getChunks: (kbId: string, docId: string) =>
    apiClient.get<Chunk[]>(`/api/v1/documents/${kbId}/${docId}/chunks`),

  getJobStatus: (kbId: string, docId: string) =>
    apiClient.get<IngestionJob>(`/api/v1/documents/${kbId}/${docId}/status`),

  reindex: (kbId: string, docId: string) =>
    apiClient.post<void>(`/api/v1/documents/${kbId}/${docId}/reindex`, {}),

  uploadDocument: async (kbId: string, file: File, token: string) => {
    const formData = new FormData()
    formData.append('file', file)
    const res = await fetch(
      `${process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000'}/api/v1/documents/upload/${kbId}`,
      { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: formData }
    )
    const json = await res.json()
    if (!res.ok) throw new Error(json.error?.message ?? 'Upload failed')
    return json
  },
}
